tempesta_fw/hpack.c

```diff
@@ -0,0 +1,3513 @@
+	       __func__, tbl->curr, tbl->n, tbl->length);
+
+	return 0;
+out:
+	T_DBG3("%s: item not added (r=%d)\n", __func__, r);
+	tfw_hpack_free_entry(&entry);
+
+	return r;
+}
+
+static const TfwHPackEntry *
+tfw_hpack_find_index(TfwHPackDTbl *__restrict tbl, unsigned long index)
+{
+	const TfwHPackEntry *entry = NULL;
+
+	if (index <= HPACK_STATIC_ENTRIES) {
+		entry = static_table + index - 1;
+		BUG_ON(!entry->name);
+		return entry;
+	}
+
+	WARN_ON_ONCE(tbl->n > tbl->length);
+	index -= HPACK_STATIC_ENTRIES;
+	if (index <= tbl->n) {
+		unsigned int curr = tbl->curr;
+
+		if (index <= curr) {
+			curr -= index;
+		} else {
+			curr += tbl->length - index;
+		}
+		T_DBG3("%s: tbl->length=%u, tbl->curr=%u, curr=%u, index=%lu\n",
+		      __func__, tbl->length, tbl->curr, curr, index);
+		entry = tbl->entries + curr;
+		BUG_ON(!entry->name || !entry->value);
+		return entry;
+	}
+
+	return NULL;
+}
+
+static int
+tfw_hpack_set_length(TfwHPack *__restrict hp, unsigned long new_size)
+{
+	TfwHPackDTbl *tbl = &hp->dec_tbl;
+	unsigned int size = tbl->size;
+
+	if (new_size > hp->max_window)
+		return -EINVAL;
+
+	if (size > new_size) {
+		unsigned int count = tbl->n;
+		unsigned int early = tbl->curr;
+		const unsigned int length = tbl->length;
+		TfwHPackEntry *const entries = tbl->entries;
+		TfwHPackEntry *cp;
+
+		if (early >= count) {
+			early -= count;
+		} else {
+			early += length - count;
+		}
+		T_DBG3("%s: tbl->curr=%u, early=%u, count=%u, length=%u,"
+		       " new_size=%lu\n", __func__, tbl->curr, early, count,
+		       length, new_size);
+		cp = entries + early;
+		do {
+			TfwHPackStr *np = cp->name;
+			TfwHPackStr *vp = cp->value;
+
+			WARN_ON_ONCE(!np->len);
+			WARN_ON_ONCE(!vp->len);
+			size -= HPACK_ENTRY_OVERHEAD + np->len + vp->len;
+
+			T_DBG3("%s: drop index, early=%u, count=%u,"
+			       " length=%u\n", __func__, early, count, length);
+			tfw_hpack_free_entry(cp);
+			early++;
+			cp++;
+			count--;
+			if (unlikely(early == length)) {
+				early = 0;
+				cp = entries;
+			}
+		} while (size > new_size);
+
+		tbl->n = count;
+		tbl->size = size;
+	}
+
+	tbl->window = new_size;
+
+	return 0;
+}
+
+static inline void
+tfw_huffman_init(TfwHPack *__restrict hp)
+{
+	hp->curr = -HT_NBITS;
+}
+
+int
+tfw_hpack_init(TfwHPack *__restrict hp, unsigned int htbl_sz)
+{
+	bool np;
+	TfwHPackETbl *et = &hp->enc_tbl;
+	TfwHPackDTbl *dt = &hp->dec_tbl;
+
+	BUILD_BUG_ON(sizeof(TfwHPackNode) > HPACK_ENTRY_OVERHEAD
+		     || HPACK_ENC_TABLE_MAX_SIZE > SHRT_MAX);
+
+	tfw_huffman_init(hp);
+
+	dt->window = hp->max_window = htbl_sz;
+	if (!(dt->pool = __tfw_pool_new(0)))
+		return -ENOMEM;
+
+	et->window = htbl_sz;
+	spin_lock_init(&et->lock);
+	et->rb_size = HPACK_ENC_TABLE_MAX_SIZE;
+	if (!(et->pool = __tfw_pool_new(HPACK_ENC_TABLE_MAX_SIZE))) {
+		tfw_pool_destroy(dt->pool);
+		return -ENOMEM;
+	}
+	et->rbuf = __tfw_pool_alloc(et->pool, HPACK_ENC_TABLE_MAX_SIZE,
+				    true, &np);
+	BUG_ON(np || !et->rbuf);
```

Comment:
`__tfw_pool_alloc()` sets `*new_page = false` and next we can fail on `__get_free_pages()` - why to crash?

Reply:
Since new allocations should not occur during `__tfw_pool_alloc()` call - BUG_ON on `np` checks that it is true (no new page, so no new allocations). But if the allocation erroneously takes place and new page is not allocated due to `-ENOMEM` - BUG_ON on `np` will not catch that case, and that's why additional BUG_ON on `!et->rbuf` is introduced.

Reply:
Oh, got it. Please comment the assertion that we allocate the pool with the maximum allowed size for the dynamic table, so the poll never grows.